File: typescript/src/perfect-numbers.ts
export function classify(n: number) {
  const range = function(startAt:number, endAt:number) {
    if (startAt > endAt) throw new Error("startAt must not be larger than endAt.")
    return Array.from({length: endAt - startAt + 1}, (_, index) => index + startAt)
  }
  
  const calculateAliquot = function(n: number) {
    return n==1? 0: range(1, n - 1).reduce((a, v) => a + (n % v == 0 ? v : 0) ) 
  }

  if (n < 1) throw new Error("Classification is only possible for natural numbers.")
  
  const aliquot = calculateAliquot(n)
  return aliquot == n ? 'perfect': (aliquot > n ? 'abundant': 'deficient')
}

File: scala/src/main/scala/PerfectNumbers.scala
object NumberType extends Enumeration {
  val Perfect, Abundant, Deficient: NumberType.Value = Value
}

object PerfectNumbers {
  private def factors(num: Int) = (1 until 1 + num/2).filter(f => num % f == 0)
  private def aliquotSum(num: Int) = factors(num).sum

  def classify(num: Int): Either[String, NumberType.Value] = {
    if (num <= 0) return Left("Classification is only possible for natural numbers.")

    val aliquot = aliquotSum(num)
    if (aliquot == num) Right(NumberType.Perfect)
    else if (aliquot < num) Right(NumberType.Deficient)
    else Right(NumberType.Abundant)
  }
}
